rmw_zenoh_cpp/src/detail/rmw_client_data.cpp

```diff
@@ -471,23 +388,66 @@ rmw_ret_t ClientData::send_request(
   // Latest consolidation guarantees unicity of replies for the same key expression,
   // which optimizes bandwidth. The default is "None", which imples replies may come in any order
   // and any number.
-  opts.consolidation = z_query_consolidation_latest();
-  z_owned_bytes_t payload;
-  z_bytes_copy_from_buf(
-    &payload, reinterpret_cast<const uint8_t *>(request_bytes), data_length);
-  opts.payload = z_move(payload);
-
-  // TODO(Yadunund): Once we switch to zenoh-cpp with lambda closures,
-  // capture shared_from_this() instead of this.
-  ClientDataWrapper * wrapper = new ClientDataWrapper(shared_from_this());
-  z_owned_closure_reply_t zn_closure_reply;
-  z_closure(&zn_closure_reply, client_data_handler, client_data_drop, wrapper);
-  z_get(
-    sess_->loan(),
-    z_loan(keyexpr_), "",
-    z_move(zn_closure_reply),
-    &opts);
-
+  opts.consolidation = zenoh::ConsolidationMode::Z_CONSOLIDATION_MODE_NONE;
+
+  std::vector<uint8_t> raw_bytes(
+    reinterpret_cast<const uint8_t *>(request_bytes),
+    reinterpret_cast<const uint8_t *>(request_bytes) + data_length);
+  opts.payload = zenoh::Bytes(std::move(raw_bytes));
+
+  std::weak_ptr<rmw_zenoh_cpp::ClientData> client_data = shared_from_this();
+  zenoh::ZResult result;
+  std::string parameters;
+  context_impl->session()->get(
+    keyexpr_.value(),
+    parameters,
+    [client_data](const zenoh::Reply & reply) {
+      if (!reply.is_ok()) {
+        auto reply_err_str = reply.get_err().get_payload().as_string();
+        RMW_ZENOH_LOG_ERROR_NAMED(
+          "rmw_zenoh_cpp",
+          "z_reply_is_ok returned False Reason: %s",
+          reply_err_str.c_str())
+        return;
+      }
+      const zenoh::Sample & sample = reply.get_ok();
+
+      auto sub_data = client_data.lock();
+      if (sub_data == nullptr) {
+        RMW_ZENOH_LOG_ERROR_NAMED(
+          "rmw_zenoh_cpp",
+          "Unable to obtain ClientData from data for %s.",
+          std::string(sample.get_keyexpr().as_string_view()).c_str());
+        return;
+      }
+
+      if (sub_data->is_shutdown()) {
+        return;
+      }
+
+      std::chrono::nanoseconds::rep received_timestamp =
+      std::chrono::system_clock::now().time_since_epoch().count();
+
+      sub_data->add_new_reply(
+        std::make_unique<rmw_zenoh_cpp::ZenohReply>(reply, received_timestamp));
+    },
+    [client_data]() {
+      auto sub_data = client_data.lock();
+      if (sub_data == nullptr) {
+        RMW_ZENOH_LOG_ERROR_NAMED(
+          "rmw_zenoh_cpp",
+          "Unable to obtain ClientData");
+        return;
+      }
+    },
```

Comment:
Hm, I'm kind of thinking that we don't need a `data_drop` callback at all anymore.  Given the fact that we are now capturing the weak pointer for the data callback, we know that either the ClientData object is alive, or not.  We no longer need to do any additional tracking.  In other words, I think this can now be a "null" callback.